### PR TITLE
Avoid multiple initializations of AmdGpu

### PR DESCRIPTION
**Pull request by epinter**
ADL finds multiple logical gpus, this commit avoids multiple unneeded initializations of AmdGpu class.

### CODE REVIEW

**Comment by epinter** [comment]
I will make one more test running as a library, just to be sure is working.

**Comment by epinter** [comment]
Doesn't work as a library, it's more complicated than it looks. I will try to work on this, check if this multiple initialization impacts the sensors.

**Comment by epinter** [comment]
@PhyxionNL
I saw your work on issue #516. Now with PMLog, I think the multiple initializations may be causing trouble in some situations. I can't reproduce myself the wrong data showing in the temperature sensors some users are reporting (on fancontrol issue), but today I saw some weird behaviour when I enable amd overlay, the temperature disappear for a second (replaced by a "-"), then shows again, by my tests always with amd overlay enabled. I think the old ADL initialization can cause this, together with the multiple initialization of PMlog.
In this commit I moved the ADL main control to amdgpugroup and replaced the ADL_* methods with ADL2_*. In my tests, I don't see the temperature disappear anymore. What you think ?

**Comment by PhyxionNL** [comment]
> @PhyxionNL I saw your work on issue #516. Now with PMLog, I think the multiple initializations may be causing trouble in some situations. I can't reproduce myself the wrong data showing in the temperature sensors some users are reporting (on fancontrol issue), but today I saw some weird behaviour when I enable amd overlay, the temperature disappear for a second (replaced by a "-"), then shows again, by my tests always with amd overlay enabled. I think the old ADL initialization can cause this, together with the multiple initialization of PMlog. In this commit I moved the ADL main control to amdgpugroup and replaced the ADL_* methods with ADL2_*. In my tests, I don't see the temperature disappear anymore. What you think ?

I think this makes sense, I'll leave some comments in a sec.

File: LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs [hunk 1]
```diff
@@ -1,12 +1,13 @@
 // This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 // If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 // Copyright (C) LibreHardwareMonitor and Contributors.
-// Partial Copyright (C) Michael M�ller <mmoeller@openhardwaremonitor.org> and Contributors.
+// Partial Copyright (C) Michael M�ller <mmoeller@openhardwaremonitor.org> and Contributors.
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs:4]: Undo.

File: LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs [hunk 2]
```diff
@@ -45,15 +48,25 @@ public AmdGpuGroup(ISettings settings)
...
+                    if (AtiAdlxx.ADL2_Adapter_AdapterInfo_Get(ref _context, adapterInfo) == AtiAdlxx.ADLStatus.ADL_OK)
                     {
                         for (int i = 0; i < numberOfAdapters; i++)
                         {
-                            AtiAdlxx.ADL_Adapter_Active_Get(adapterInfo[i].AdapterIndex, out int isActive);
+                            uint device = 0;
+                            AtiAdlxx.ADLGcnInfo gcnInfo = new();
+                            AtiAdlxx.ADLPMLogSupportInfo pmLogSupportInfo = new();
+                            AtiAdlxx.ADL2_Adapter_Active_Get(_context, adapterInfo[i].AdapterIndex, out int isActive);
+
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs:60]: One line too much.

File: LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs [hunk 3]
```diff
@@ -77,9 +90,43 @@ public AmdGpuGroup(ISettings settings)
...
+                                _report.AppendLine("Sensors Supported: " + sensorsSupported);
+
+                                if (device != 0)
+                                {
+                                    AtiAdlxx.ADL2_Device_PMLog_Device_Destroy(_context, device);
+                                }
+                            }
+
+                            if (!string.IsNullOrEmpty(adapterInfo[i].UDID) && adapterInfo[i].VendorID == AtiAdlxx.ATI_VENDOR_ID &&
+                                !IsAdapterAlreadyFound(adapterInfo[i].BusNumber, adapterInfo[i].DeviceNumber))
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs:119]: Is this ever returning the same adapter here?

File: LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs [hunk 4]
```diff
@@ -104,6 +151,18 @@ public AmdGpuGroup(ISettings settings)
         }
     }
 
+    private bool IsAdapterAlreadyFound(int busNumber, int deviceNumber)
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs:154]: `IsAlreadyAdded` ?

**Review by PhyxionNL — CHANGES_REQUESTED** [review]

**Comment by epinter** [comment]
I didn't finished the code yet, there's unused imports, some code I need to double check :)
Now that you agree with the commit, I will review and finish it.

File: LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs (hunk 3 above)
Comment (epinter) [reply LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs:119]: Yes, in my case I have 5, all of them have all sensors:

```
Number of adapters: 9

AdapterIndex: 0
isActive: 1
AdapterName: AMD Radeon RX 7800 XT
UDID: PCI_VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8_6&8C9B63C&0&00000008A
PNPString: PCI\VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8\6&8C9B63C&0&00000008
Present: 1
VendorID: 0x1002
BusNumber: 3
DeviceNumber: 0
FunctionNumber: 0
AdapterID: 0x35CAF060

AdapterIndex: 1
isActive: 0
AdapterName: AMD Radeon RX 7800 XT
UDID: PCI_VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8_6&8C9B63C&0&00000008&02A
PNPString: PCI\VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8\6&8C9B63C&0&00000008&02
Present: 1
VendorID: 0x1002
BusNumber: 3
DeviceNumber: 0
FunctionNumber: 0
AdapterID: 0x35CAF060

AdapterIndex: 2
isActive: 0
AdapterName: AMD Radeon RX 7800 XT
UDID: PCI_VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8_6&8C9B63C&0&00000008&03A
PNPString: PCI\VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8\6&8C9B63C&0&00000008&03
Present: 1
VendorID: 0x1002
BusNumber: 3
DeviceNumber: 0
FunctionNumber: 0
AdapterID: 0x35CAF060

AdapterIndex: 3
isActive: 0
AdapterName: AMD Radeon RX 7800 XT
UDID: PCI_VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8_6&8C9B63C&0&00000008&04A
PNPString: PCI\VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8\6&8C9B63C&0&00000008&04
Present: 1
VendorID: 0x1002
BusNumber: 3
DeviceNumber: 0
FunctionNumber: 0
AdapterID: 0x35CAF060

AdapterIndex: 4
isActive: 0
AdapterName: AMD Radeon RX 7800 XT
UDID: PCI_VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8_6&8C9B63C&0&00000008&05A
PNPString: PCI\VEN_1002&DEV_747E&SUBSYS_05FD1043&REV_C8\6&8C9B63C&0&00000008&05
Present: 1
VendorID: 0x1002
BusNumber: 3
DeviceNumber: 0
FunctionNumber: 0
AdapterID: 0x35CAF060
```

File: LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs (hunk 1 above)
Comment (epinter) [reply LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs:4]: I tried, I will see how I can fix this.

Comment (epinter) [reply LibreHardwareMonitorLib/Hardware/Gpu/AmdGpuGroup.cs]: Copied the line from amdgpu.cs

**Comment by epinter** [comment]
I think it is ready.

**Comment by PhyxionNL** [comment]
> I think it is ready.

Thanks!

**Pull request merged**